DM-50562: Updates for support of staged chunk ingestion pipeline in BigQuery

JeremyMcCormick opened the pull request:
This PR updates the replication tools to support a multi-stage ingestion of APDB replica chunks into BigQuery, so that the replica chunks are first copied into staging tables before being promoted to production. The bulk of this new behavior is implemented in other repositories such as `lsst-dm/ppdb-cloud-functions`, but this repo needed to be updated so that the "PPDB" Postgres database it creates could be used for managing the status of the chunks both on-premises at the USDF and in the Google Cloud environment. So the primary instance of this Postgres will be cloud-hosted and not at the USDF. (It was deemed easier to access the cloud Postgres database from the USDF rather than vice versa.)

The primary change to support this system was switching from directory tree scanning to usage of the Postgres database for managing the status of the replica chunks and the location of their locally exported files. The updates included improvements and updates to the `ChunkExporter` and `ChunkUploader` classes. These were basically rewritten, incorporating various improvements. Functionality involving Google Cloud Platform (GCP) libraries (`google` and `google.cloud` packages) was refactored into the [dax_ppdbx_gcp](https://github.com/lsst-dm/dax_ppdbx_gcp) repository, which is now an optional dependency that can be installed using `pip install .[gcp]`. The export tools cannot be used if this package is missing from the Python environment. (An error will be thrown when importing the `lsst.dax.ppdb.export` module if it is not installed.)

**Major Changes**

- Changed the `export` modules to use a database with a modified version of the `PpdbReplicaChunk` table for management of chunks, instead of scanning the file system
  - The `ChunkUploader` which uploads the parquet files to GCS now depends on the PPDB config, as it directly updates the status of the replica chunks in the Postgres database to "uploaded."
  - The `ChunkExporter` now subclasses `PpdbReplicaChunkSql`, a new class which includes some minor extensions to `PpdbSql` for tracking and managing replica chunks.
- Moved all Google Cloud functionality into the `dax_ppdbx_gcp` repository and added it to `requirementst.txt` and as part of an optional `gcp` set of dependencies in `pyproject.yaml`
- Used the newly introduced functionality in `ApdbTableData` for getting column datatypes when exporting to parquet
- `ChunkExporter` now publishes a Pub/Sub event to update the status of replica chunks to "exported" instead of directly connecting to the Postgres database.

**Minor Changes**

- Added a `--delete-existing` flag to the chunk exporter, mainly for use in development and testing
  - If a directory is found for a replica chunk when this flag is disabled, an error will now be thrown.
- Removed the `--delete-chunks` flag from the `ChunkUploader`, as it shouldn't ever be deleting any exported parquet files
- Generated a more complete and useful manifest file for each replica chunk when exporting to parquet
  - This manifest is modified with an upload date field and then copied directly into GCS when uploading the parquet files. 
- Added a few optional fields to the `ppdb.PpdbReplicaChunk` interface to support tracking replica chunk status using the local "PPDB" Postgres database
  - These included a `status` field (exported, uploaded, etc.) and `directory`, which points to the path where the replica chunk's parquet files and manifest have been stored locally. 
- Modified the timers in the `export` modules to use an APDB monitoring agent rather than write directly to the log (should be useful for making Grafana dashboards in the future)
- Completely revised error handling and control flow in the chunk uploader and exporter
  - Added module-specific error types
  - Made a best effort attempt to handle various error states that might occur
- Fixed mypy errors in `_chunk_uploader` and removed its ignore from `mypy.ini`
- Removed Python 3.10 from the supported versions listed in the project classifiers and added Python 3.13 support
  - Python 3.10 was not present in the existing build matrix; Python 3.13 was added.
- Moved Arrow functions for exporting parquet data into a separate `_arrow` module
  - Revised and refactored the parquet functions in the process of doing this.
- Removed the  `lastNonForcedSource` column from test schema, as its presence was causing local test failures.
- Refactored the `ppdb_sql` module so that the `PpdbReplicaChunk` table is created in a separate method, making it easier to add additional columns to this table in a child class
- Added the standard DM `docstyle` workflow along with project configuration for `numpydoc` and `pydocstyle` (copied from `daf_butler`)

**Notes**

- The new `_ppdb_replica_chunk_sql` module extends the functionality of `_ppdb_sql` to support additional database columns in the `PpdbReplicaChunk` table. This is not intended to be a final implementation. Revising the `Ppdb` and `PpdbSql` classes to separate out the replica chunk management from the actual PPDB database is a non-trivial task, which may require extensive revision of existing classes and interfaces, so this work has been moved to a separate ticket ([DM-52173](https://rubinobs.atlassian.net/browse/DM-52173)).
- Setting of the `validityEnd` dates in the `DiaObject` table were _not_ implemented on this ticket branch. This work will be done on [DM-50641](https://rubinobs.atlassian.net/browse/DM-50641).
- Test coverage for these changes and the `export` module in general is basically non-existent. I have been testing manually with external scripts. This lack of test cases can be addressed on a separate ticket, after discussion of how this should be done w.r.t. the cloud environment setup and mocking. Since this module uses a number of Google Cloud services, as well as a Postgres database, setting up useful test cases will not be very straightforward. (For one, valid Google Cloud credentials would need to be provided when testing if the services were not mocked locally.)

[DM-52173]: https://rubinobs.atlassian.net/browse/DM-52173?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**TODO**

- [x] Fix [failing tests](https://github.com/lsst/dax_ppdb/actions/runs/17050745655/job/48337499632?pr=14), likely due to updates in `dax_apdb`
- [ ] Update `dax_ppdbx_gcp` dependency to `main` after https://github.com/lsst-dm/dax_ppdbx_gcp/pull/1 is merged (currently depending on a release tag)

**Changes Requested**

These are [changes requested in review](https://github.com/lsst/dax_ppdb/pull/14#pullrequestreview-3137775038) from @andy-slac.

- [x] `ChunkExporter` needs separate config class which keeps configurable parameters like `batch_size` or `topic_name`.
- [x] `_factory` module needs to be extended to handle ChunkExporter's config class, so that `Ppdb.from_uri` returns `ChunkExporter` instance when it reads its config. In that way you won't need a separate `export_chunks_run` script, `replication_run` will handle all implementations.
- [x] Command is needed for creating the Postgres database with extra `PpdbReplicaChunk` fields using in `_export` module.

It was also suggested to refactor `ChunkExporter` so that it does not inherit from `PpdbSql`. There were also questions about whether the `PpdbReplicaChunkSql` class should exist at all. I propose to move that work to [DM-52173](https://rubinobs.atlassian.net/browse/DM-52173) instead, since it will require some fairly extensive refactoring of the current class structure.

JeremyMcCormick committed:
Write a sidecar metadata file for the chunk after exporting

The metadata includes the replica chunk id, the unique id of the chunk,
the APDB schema version, export date, last update time of the chunk,
table names and their row counts, and the compression format used in
writing to Parquet.

JeremyMcCormick committed:
Do not include the schema version in the local chunk dir

This may be included in the full object name in cloud storage to
facilitate searching, but it should not be necessary for managing the
exported chunks locally.

JeremyMcCormick committed:
Add a status field to the metadata for management of uploads

JeremyMcCormick committed:
Use the TAI time from the replica chunk instead of formatting in ISO

JeremyMcCormick committed:
Add a status column to the PpdbReplicaChunk table

JeremyMcCormick committed:
Add a status field to the PpdbReplicaChunk table

JeremyMcCormick committed:
Add a directory column to the PpdbReplicaChunk table

This column points to the local directory where the replica chunk files
have been exported.

JeremyMcCormick committed:
Add option to query the PpdbReplicaChunk table by chunk status

JeremyMcCormick committed:
Rearchitect the chunk uploader to work from the database

This changes the ``ChunkUploader`` to work from the PPDB database for
determining which chunks can be uploaded to cloud storage.

JeremyMcCormick committed:
Normalize the name of the manifest file

This changes the local manifest file to use ".manifest" rather than
".metadata" to match the file name from cloud storage.

JeremyMcCormick committed:
Remove option from chunk uploader for automatically deleting files

We don't ever want the uploader to do this automatically.

JeremyMcCormick committed:
Fix issue with nested fstrings

JeremyMcCormick committed:
Remove outdated method

JeremyMcCormick committed:
Add the chunk ID to the Pub/Sub event payload

Also change the bucket and path to just a full GCS URL pointing to the
prefix with the chunk files.

JeremyMcCormick committed:
Add GCP auth helper function

JeremyMcCormick committed:
Add helper function for publishing a message to pubsub

JeremyMcCormick committed:
Publish a pubsub event to insert a new chunk in the tracking database

Instead of directly connecting to the Postgres database, the chunk
exporter now publishes a message to Pub/Sub.

JeremyMcCormick committed:
Make the pubsub topic configurable

JeremyMcCormick committed:
Add method for validating that the pubsub topic exists

JeremyMcCormick committed:
Use utility methods for GCP auth and Pub/Sub publishing

JeremyMcCormick committed:
Add an `--overwrite` flag for enabling overwrite of existing chunk data

By default, the chunk exporter will fail if a directory already exists
for the chunk. This flag can be used to override this behavior and
replace the existing directory. For the production system, it is likely
that existing directories should not be overwritten, because this could
lead to duplicate data.

JeremyMcCormick committed:
Change `overwrite` option of `export-chunks` to `delete-existing`

JeremyMcCormick committed:
Use ppdb-gcp repository for Google Cloud functionality

This removes several GCP utility modules which were ported to the
ppdb-gcp repository. It also updates the project requirements to
include ppdb-gcp and fixes up the Python imports. The classes being
used previously are currently the same in the new repo so no
additional changes were needed.

JeremyMcCormick committed:
Introduce PpdbReplicaChunkSql placeholder class and revert changes

This reverts the changes to `ppdb.sql` and `_ppdb_sql.py` and
introduces a placeholder `PpdbReplicaChunkSql` class to provide
replica chunk management functionality without breaking backward
compatibility.

JeremyMcCormick committed:
Revise chunk exporter to use new ApdbTableData interface

Instead of getting the column datatypes from the schema, the chunk
exporter will now read them directly from the `ApdbTableData` using the
`column_defs` method. A small module was added which provides the
mapping from the Felis `DataType` to arrow types, as well as a simple
function for creating an Arrow schema.

JeremyMcCormick committed:
Remove the chunk status from the manifest file

The chunk tracking database is the canonical source of information for
the status of a chunk rather than the manifest.

JeremyMcCormick committed:
Exclude several unwanted APDB columns from Parquet output

JeremyMcCormick committed:
Move method for writing parquet files into utility module

JeremyMcCormick committed:
Move the arrow util module one level up

JeremyMcCormick committed:
Make the Google Cloud Platform (GCP) dependencies optional

The GCP dependencies are encapsulated in a separate project, which can
be installed using the `gcp` optional dependency set.

When importing the `lsst.dax.ppdb.export` module, an error will be
raised if these dependencies are missing. The chunk exporter and
uploader cannot be run without them.

JeremyMcCormick committed:
Fix mypy errors and various typing issues

JeremyMcCormick committed:
Cleanup modifications to ppdb interfaces

JeremyMcCormick committed:
Remove status from manifest

JeremyMcCormick committed:
Add optional depedency to build

JeremyMcCormick committed:
Update project and package name of GCP utility package

The repository was renamed to `dax_ppdbx_gcp` and the package is now
`lsst.dax.ppdbx.gcp`.

JeremyMcCormick committed:
Fix typing errors in `_chunk_uploader`

JeremyMcCormick committed:
Refactor methods in `_ppdb_sql` to reduce duplication in child class

JeremyMcCormick committed:
Use APDB monitoring so that extra values will show in print outs

JeremyMcCormick committed:
Add `google-cloud-storage` to `gcp` optional dependencies

This API is used directly in the chunk uploader and so should be
installed explicitly even though it is a dependency of
`lsst-dax-ppdbx-gcp`.

JeremyMcCormick committed:
Fix bug in defining dictionary of table data for replication

JeremyMcCormick committed:
Add dax_ppdbx_gcp requirement

JeremyMcCormick committed:
Add Python 3.13 to build matrix

JeremyMcCormick committed:
Refactor to use storage client from `lsst.dax.ppdbx.gcp.gcs` module

JeremyMcCormick committed:
Remove Python 3.10 classifier and add 3.13

JeremyMcCormick committed:
Revise error handling

JeremyMcCormick committed:
Add docstyle workflow

JeremyMcCormick committed:
Add `tool.numpydoc_validation` section

JeremyMcCormick committed:
Add `tool.pydocstyle` section

This was copied from `daf_butler`.

JeremyMcCormick committed:
Fix missing docstrings reported by pydocstyle

andy-slac reviewed (changes requested):
OK, this is not going in the direction that I expected, and it becomes quite messy. I think we need to work out a few things before we can merge this:

- I see no reason for `PpdbReplicaChunkSql` to exist. There should be two implementations of `Ppdb` interface, one that is current `PpdbSql` without any extensions, another one is for use with replication to GCS, which is basically current `ChunkExporter`. `ChunkExporter` should not be inheriting `PpdbSql`.
- I do not see how do you create database for use with GCS replication, there is no CLI for that?
- `ChunkExporter` (and we should probably rename that) needs it separate config class which keeps configurable parameters like `batch_size` or `topic_name`.
- `_factory` module needs to be extended to handle ChunkExporter's config class, so that `Ppdb.from_uri` returns `ChunkExporter` instance when it reads its config. In that way you won't need a separate `export_chunks_run` script, `replication_run` will handle all implementations.

andy-slac commented:
> I can make that change. The new exporter implementing the `Ppdb` interface will have some duplication with `PpdbSql` though, unless there is a common class with the `PpdbReplicaChunk` functionality.

I do not think there is much duplication, the exporter should only care about a single table for replica chunks, the rest of PpdbSql is not at all useful for you, and it actually creates problems w.r.t. schema version management (it has to be done differently in exporter).

> I have just been using the full "PPDB" database created with the existing CLI tools 

I do not see how. Those tools would create `PpdbReplicaChunks` table without extra columns that exporter needs.

> Not sure I understand this part. Can you outline this with a bit more detail about how this would work?

The only difference between `export_chunks_run` and `replication_run` is the type of the instance of `Ppdb` that they create. If that difference can be hidden in `Ppdb.from_uri` then `replication_run` can handle both cases.

JeremyMcCormick commented on python/lsst/dax/ppdb/export/_chunk_exporter.py:
```diff
@@ -119,100 +144,110 @@ def store(
+    ) -> None:
+        """Publish a message to the 'track-chunk-topic' Pub/Sub topic.
+
+        This will add a new record to the chunk tracking database with status
+        of 'exported'. The chunk uploader process will then pick up this record
+        and copy the chunk into cloud storage.
```
"Chunk tracking database" -> PpdbReplicaChunk table.

andy-slac replied:
Why do you need a cloud function to update PpdbReplicaChunk table? You can do it right here, without extra dependency (and potential failure mode) on cloud.

JeremyMcCormick commented on python/lsst/dax/ppdb/tests/_ppdb.py:
```diff
@@ -115,7 +116,9 @@ def _fill_apdb(self, apdb: Apdb) -> None:
             apdb.store(visit_time, objects, sources, fsources)
             start_id += nobj
 
-    def _check_chunks(self, apdb_chunks: list[ReplicaChunk], ppdb_chunks: list[PpdbReplicaChunk]) -> None:
+    def _check_chunks(
+        self, apdb_chunks: Sequence[ReplicaChunk], ppdb_chunks: Sequence[PpdbReplicaChunk]
+    ) -> None:
```
Mistake - I will revert.

andy-slac commented on python/lsst/dax/ppdb/export/_chunk_uploader.py:
```diff
@@ -19,417 +19,305 @@
 
-    def _upload_single_file(self, file_path: Path, gcs_name: str) -> None:
-        """Upload a single file to GCS.
+        manifest_path = chunk_dir / f"chunk_{chunk_id}.manifest.json"
```
I mean that you evaluate this path in more than one place, to keep it consistent do it in one place.

JeremyMcCormick committed:
Use ticket branch for dax_ppdbx_gcp dependency

JeremyMcCormick committed:
Improve lookup of arrow type

JeremyMcCormick committed:
Change mapping of Felis float type to float32 in pyarrow from float64

JeremyMcCormick committed:
Fix parameter referenes in docstrings

JeremyMcCormick committed:
Remove unnecessary int conversion on batch_size

JeremyMcCormick commented on python/lsst/dax/ppdb/cli/options.py:
```diff
@@ -119,6 +119,12 @@ def replication_options(parser: argparse.ArgumentParser) -> None:
 def export_options(parser: argparse.ArgumentParser) -> None:
     group = parser.add_argument_group("chunk export options")
     group.add_argument("--directory", help="Directory for local file storage.", required=True)
+    group.add_argument(
+        "--topic",
+        help="Pub/Sub topic for publishing the exported replica chunks.",
+        default=None,
+        required=False,
+    )
```
Yes, I think your idea about moving the command line options into a config class for `ChunkExporter` is a good one. I will do that after all the other basic cleanup you requested.

JeremyMcCormick committed:
Remove unnecessary directory check

JeremyMcCormick committed:
Correct typing of manifest_data function parameter

JeremyMcCormick committed:
Remove unnecessary string conversion

JeremyMcCormick committed:
Correct variable name in exception logging

JeremyMcCormick committed:
Update class notes in docstrings

JeremyMcCormick committed:
Use `sys.exit` instead of `SystemExit`

JeremyMcCormick commented on python/lsst/dax/ppdb/export/_chunk_uploader.py:
```diff
@@ -19,417 +19,305 @@
+                    "upload_files_time", _MON, tags={"prefix": str(gcs_prefix), "chunk_id": str(chunk_id)}
+                ) as timer:
+                    self.storage.upload_files(gcs_names)
+                    timer.add_values(file_count=len(gcs_names))
```
The timer measures how long it takes to upload all the files in the chunk, because `upload_files` uses a thread pool executor to upload them in parallel. If `add_values` can accept a dictionary argument, I could provide a mapping of filename -> size.

JeremyMcCormick committed:
Revert unwanted typing change

JeremyMcCormick commented on python/lsst/dax/ppdb/export/_chunk_exporter.py:
```diff
@@ -73,41 +61,78 @@ class ChunkExporter(PpdbSql):
+
+        self.delete_existing = delete_existing
+
+    def _generate_manifest_data(
+        self, replica_chunk: ReplicaChunk, table_dict: dict[str, ApdbTableData]
+    ) -> dict[str, Any]:
+        """Generate the manifest data for the replica chunk."""
+        return {
+            "chunk_id": str(replica_chunk.id),
+            "unique_id": str(replica_chunk.unique_id),
+            "schema_version": str(self.schema_version),
+            "exported_at": datetime.now(timezone.utc).isoformat(),
+            "last_update_time": str(replica_chunk.last_update_time),  # TAI value
+            "table_data": {
+                table_name: {
+                    "row_count": len(data.rows()),
+                }
+                for table_name, data in table_dict.items()
+            },
+            "compression_format": self.compression_format,
+        }
```
> I think it will help if you make pydantic model for the manifest.

Did you actually mean a Python `dataclass`? 

Pydantic is available through a transitive dependency because Felis depends on it. And I do prefer working with Pydantic over `dataclasses`, but I just wanted to make sure it would be okay to introduce that dependency here. 

Probably `pydantic` should also be added to project deps if importing it directly.

timj replied:
You are encouraged to use pydantic and yes you should never rely on a dependency you are using directly to be pulled in from another dependency so it would need to be added .

andy-slac replied:
Pydantic is indeed already a dependency but is missing from requirements, has to be added.

JeremyMcCormick commented on python/lsst/dax/ppdb/sql/_ppdb_replica_chunk_sql.py:
```diff
@@ -0,0 +1,163 @@
+"""
+
+
+class PpdbReplicaChunkSql(PpdbSql):
```
How about something like `ParquetExtractor` since that's what it is doing? (Open to other suggestions you might have.)

I'm also inclined to rename the `export` package itself to `etl`, as that's really what its modules are implementing (Extract, Transform, Load operations for APDB -> PPDB).

andy-slac replied:
I like `etl` even less than `extractor`, there is no transform happening, PPDB is a pure one-to-one _replica_. And I said already that `PpdbReplicaChunkSql` should not even need to exist.

JeremyMcCormick replied:
ETL is industry-standard terminology using to describe processes that load data from one source into another. It doesn't necessarily imply that a "transform" is always applied. 

The name of the package doesn't really matter that much to me though. I can leave as is for now.

JeremyMcCormick replied:
Discussed in PR description - we will do this refactoring and reorganization on DM-52173.

JeremyMcCormick committed:
Add pydantic dependency

JeremyMcCormick committed:
Introduce a pydantic model for representing the replica chunk manifest

JeremyMcCormick committed:
Add total bytes uploaded to timer values

JeremyMcCormick committed:
Do not exclude apdb_replica_chunk field when exporting to parquet

JeremyMcCormick committed:
Move the configuration for `ChunkExporter` into `PpdbConfig` class

This changes how the `ChunkExporter` is configured and instantiated.
It is now created similarly to `PpdbSql` and does not need its own
replication sub-command any longer.

JeremyMcCormick committed:
Make method for storing chunk data public in parent class

JeremyMcCormick committed:
Get schema version from metadata

JeremyMcCormick committed:
Remove leading underscore from a few methods in `PpdbSql`

JeremyMcCormick committed:
Move some parameters for chunk uploader into config class

This moves the parameters for the GCS bucket name, GCS prefix, and
BigQuery dataset name into the `PpdbBigQueryConfig` class. These were
removed from the command line interface for the uploader.

JeremyMcCormick committed:
Fix module name for `PpdbBigQueryConfig`

JeremyMcCormick committed:
Cleanup modifications to `Ppdb` and `PpdbSql` classes

This cleans up some of the modifications of the `Ppdb` and `PpdbSql`
classes so they do not include extra replica chunk information used in
the `export` module.

JeremyMcCormick committed:
Add replica chunk directly to database without Pub/Sub event

JeremyMcCormick committed:
Move name of Pub/Sub topic name for triggering chunk staging to config

JeremyMcCormick commented:
@andy-slac The `export` package could be renamed to `bigquery`, since all of its modules related to the PPDB BigQuery implementation.

JeremyMcCormick commented:
@andy-slac 

> I do not see how do you create database for use with GCS replication, there is no CLI for that?

Do you have any preference how this is done?

For the full PPDB SQL database, I see that `ppdb_cli` calls `scripts/create_sql`. Should I add an analogous module for the "bigquery" implementation and add a command for it to the CLI?

Whatever I do is probably going to be reworked in DM-52173, so this is likely to just be an interim solution.

JeremyMcCormick committed:
Rename `export` package to `bigquery`

This package contains tools for replicating data to BigQuery.

JeremyMcCormick committed:
Add `create-bq-replica-chunk-sql` command to `ppdb-cli`

This adds a command to create the Postgres database used to track
replica chunk information for BigQuery.

JeremyMcCormick committed:
Fix minor docstring typing error and a few comment typos

andy-slac reviewed (approved):
Looks OK, a small bunch of comments and docstring fixes.

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_config.py:
```diff
@@ -0,0 +1,37 @@
+    stage_chunk_topic: str = "stage-chunk-topic"
+    """Pub/Sub topic name for triggering chunk staging process."""
+
+    batch_size: int = 1000
+    """Number of rows to process in each batch when writing parquet files."""
```
Noted on [DM-52326](https://rubinobs.atlassian.net/browse/DM-52326).

JeremyMcCormick commented on python/lsst/dax/ppdb/sql/_ppdb_sql.py:
```diff
@@ -501,22 +520,29 @@ def store(
 
-    def _store_insert_id(
-        self, replica_chunk: ReplicaChunk, connection: sqlalchemy.engine.Connection, update: bool
+    def upsert(
```
I'll refactor the store and upsert methods to make more sense.

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_chunk_exporter.py:
```diff
@@ -0,0 +1,202 @@
+        try:
+            chunk_dir = self._get_chunk_path(replica_chunk.id)
+
+            if chunk_dir.exists():
```
Noted in DM-52326.

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_chunk_uploader.py:
```diff
@@ -0,0 +1,296 @@
+        self._sql = PpdbReplicaChunkSql(config)
+
+        # Read parameters from config
+        if self.config.prefix is None:
+            raise ValueError("GCS prefix is not set in configuration.")
```
For now, I will leave these checks, because `create-bigquery-replica-chunk-sql` doesn't include any of these optional fields when generating the config, So I don't want to make it part of the validation right now or it won't load. 

I will make a note to improve this Pydantic model in DM-52326. I'm also planning to add a command for generating the complete `PpdbBigQueryConfig` as well.

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_chunk_uploader.py:
```diff
@@ -0,0 +1,296 @@
+        )
+
+        # Make a list of local parquet files to upload.
+        parquet_files = list(chunk_dir.glob("*.parquet"))
```
I don't want to rearchitect this on the ticket branch, but I will make a note of it on DM-52326. I agree with you that this approach would be better than using a file glob.

JeremyMcCormick committed:
Fix minor issues with docstrings, typing and comments

JeremyMcCormick committed:
Change property names on the `PpdbReplicaChunkExtended` class

This changes the `file_name` property to `manifest_name` and
`file_path` to `manifest_path`.

JeremyMcCormick committed:
Load config using `PpdbConfig` class

JeremyMcCormick committed:
Check for correct config type before initializing parent class

JeremyMcCormick committed:
Remove unnecessary check of `compression_format`

This has a default in the Pydantic data model and so does not need to
be checked here for `None`.

JeremyMcCormick committed:
Remove `exist_ok=True` when making the chunk dir

It should already have been deleted if it exists or an error was raised
in the case the `delete_existing` flag was not enabled.

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_chunk_exporter.py:
```diff
@@ -0,0 +1,202 @@
+            # Loop over the table data and write each table to a Parquet file.
+            for table_name, table_data in table_dict.items():
+                if not table_data.rows():
+                    _LOG.warning("No data for %s in %s, skipping export", table_name, replica_chunk.id)
```
I will change to debug level.

JeremyMcCormick committed:
Change log message to debug level when there is no table data

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_chunk_exporter.py:
```diff
@@ -0,0 +1,202 @@
+            _LOG.exception("Failed to store replica chunk: %s", replica_chunk.id)
+            raise
+
+        # Store the replica chunk info in the database, including status and
+        # directory.
+        replica_chunk_ext = PpdbReplicaChunkExtended.from_replica_chunk(
+            replica_chunk, ChunkStatus.EXPORTED, chunk_dir
+        )
```
Currently, the only real recording of errors is via log messages. I made a note on DM-52326 to review and improve the error handling in both the exporter and uploader tools. Specifically, I will note the requirement to skip processing of chunks where there was a previous failure.

JeremyMcCormick committed:
Restore original version of `_store_chunk_id` method

This backs out of some changes to the `_ppdb_sql` module and instead
uses some duplicate code in `_replica_chunk` for chunk inserts to avoid
unnecessary interface changes. The code duplication will be dealt with
on DM-52173.

JeremyMcCormick commented:
> Looks OK, a small bunch of comments and docstring fixes.

These should all be either directly resolved on this ticket now, or they were added to other tickets like [DM-52326](https://rubinobs.atlassian.net/browse/DM-52326).

JeremyMcCormick commented on python/lsst/dax/ppdb/bigquery/_manifest.py:
```diff
@@ -0,0 +1,122 @@
+    compression_format
+        Name of the compression format used for artifacts (e.g., "gzip",
+        "zstd", "snappy", etc.).
+    """
```
> https://developer.lsst.io/python/numpydoc.html#placement-of-constant-and-class-attribute-docstrings

Seems like both are valid - if the inlined placement is DM-standard, I can change it.

JeremyMcCormick committed:
Put class attribute doc below declarations

https://developer.lsst.io/python/numpydoc.html#placement-of-constant-and-class-attribute-docstrings

JeremyMcCormick committed:
Switch the lsst-dax-ppdbx-gcp requirement to `main` from ticket branch

JeremyMcCormick committed:
Fix up python build
